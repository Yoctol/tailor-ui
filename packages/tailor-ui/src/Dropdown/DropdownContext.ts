import { createContext } from 'react';

const DropdownContext = createContext({
  close: () => {},
});

export default DropdownContext;
