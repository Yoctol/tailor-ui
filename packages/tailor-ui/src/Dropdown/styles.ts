import styled, { DefaultTheme, StyledComponent } from 'styled-components';
import { ColorProps, color } from 'styled-system';

export const List = styled.div`
  max-height: 240px;
  padding: ${(p) => p.theme.space[2]} 0;
  overflow-x: hidden;
  overflow-y: auto;
  border-radius: ${(p) => p.theme.radii.lg};
`;

export const Item: StyledComponent<
  'li',
  DefaultTheme,
  {
    disabled: boolean;
  } & ColorProps,
  never
> = styled.li<{ disabled: boolean } & ColorProps>`
  display: flex;
  align-items: center;
  height: ${(p) => p.theme.heights.base};
  padding: 0 24px;
  background-color: ${(p) => p.theme.colors.light};
  font-size: ${(p) => p.theme.fontSizes.sm};
  cursor: pointer;

  &[disabled],
  &:disabled {
    color: ${(p) => p.theme.colors.gray500};
    cursor: not-allowed;

    :hover {
      background-color: ${(p) => p.theme.colors.light};
    }
  }

  &:hover {
    background-color: ${(p) => p.theme.colors.gray200};
  }

  ${color};
`;

export const Divider = styled.div`
  width: 100%;
  height: 1px;
  margin: ${(p) => p.theme.space[1]} 0;
  background-color: ${(p) => p.theme.colors.gray300};
`;
