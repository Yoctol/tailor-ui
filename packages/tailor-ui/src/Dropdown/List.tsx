import React, { PureComponent } from 'react';
import styled from 'styled-components';
import {
  MaxHeightProps,
  MinWidthProps,
  OverflowProps,
  TextAlignProps,
  maxHeight,
  minWidth,
  overflow,
  textAlign,
} from 'styled-system';
import { animated } from 'react-spring/renderprops.cjs';

import { Consumer } from './DropdownContext';

export type StyledListProps = MinWidthProps &
  MaxHeightProps &
  TextAlignProps &
  OverflowProps;

export const StyledList = styled.ul<StyledListProps>`
  display: block;
  margin: 0;
  padding: ${p => p.theme.space[1]} 0;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray300};
  background-color: ${p => p.theme.colors.light};
  box-shadow: ${p => p.theme.shadows.lg};
  list-style: none;

  &:focus {
    outline: 0;
  }

  ${overflow};
  ${minWidth};
  ${maxHeight};
  ${textAlign};
`;

const AnimatedStyledList = animated(StyledList);

export type ListProps = MinWidthProps &
  TextAlignProps & {
    style?: any;
  };

class List extends PureComponent<ListProps> {
  static defaultProps = {
    textAlign: 'left',
    minWidth: 100,
  };

  render() {
    const { children, style, ...props } = this.props;
    return (
      <Consumer>
        {({ styles, handleListRef }) => (
          <AnimatedStyledList
            data-testid="dropdown-overlay"
            ref={handleListRef}
            style={{
              ...style,
              ...styles,
            }}
            {...props}
          >
            {children}
          </AnimatedStyledList>
        )}
      </Consumer>
    );
  }
}

export default List;
