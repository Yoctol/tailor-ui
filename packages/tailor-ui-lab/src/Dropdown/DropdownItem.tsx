import React, {
  FunctionComponent,
  MouseEvent,
  MouseEventHandler,
  useContext,
} from 'react';

import DropdownContext from './DropdownContext';
import { Item } from './styles';

interface DropdownItemProps {
  /**
   * Disabled the item
   */
  disabled?: boolean;
  /**
   * Callback for click
   */
  onClick?: MouseEventHandler;
  /**
   * Keep the dropdown after click item
   */
  keep?: boolean;
}

const DropdownItem: FunctionComponent<DropdownItemProps> = ({
  keep = false,
  disabled = false,
  onClick,
  ...otherProps
}) => {
  const { close } = useContext(DropdownContext);

  const handleClick = (event: MouseEvent) => {
    if (disabled) {
      return;
    }

    if (onClick) {
      onClick(event);
    }

    if (!keep) {
      close();
    }
  };

  return <Item onClick={handleClick} disabled={disabled} {...otherProps} />;
};

export default DropdownItem;
