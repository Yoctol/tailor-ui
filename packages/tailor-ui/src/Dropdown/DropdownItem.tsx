import React, {
  FC,
  MouseEvent,
  MouseEventHandler,
  useCallback,
  useContext,
} from 'react';

import DropdownContext from './DropdownContext';
import { Item } from './styles';

interface DropdownItemProps {
  /**
   * Disabled the item
   */
  disabled?: boolean;
  /**
   * Callback for click
   */
  onClick?: MouseEventHandler;
  /**
   * Keep the dropdown after click item
   */
  keep?: boolean;
}

const DropdownItem: FC<DropdownItemProps> = ({
  keep = false,
  disabled = false,
  onClick,
  ...otherProps
}) => {
  const { close } = useContext(DropdownContext);

  const handleClick = useCallback(
    (event: MouseEvent) => {
      if (disabled) {
        return;
      }

      if (onClick) {
        onClick(event);
      }

      if (!keep) {
        close();
      }
    },
    [close, disabled, keep, onClick]
  );

  return <Item onClick={handleClick} disabled={disabled} {...otherProps} />;
};

export default DropdownItem;
