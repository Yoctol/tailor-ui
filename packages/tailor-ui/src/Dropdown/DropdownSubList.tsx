import React, { FC, useState } from 'react';
import { MdKeyboardArrowRight } from 'react-icons/md';

import { Box } from '../Layout';
import { Icon } from '../Icon';
import { Position } from '../constants';
import { StyledPopover } from '../Popover/styles';
import { Tooltip } from '../Tooltip';

import { Item, List } from './styles';

interface DropdownSubListProps {
  title: string;
  disabled?: boolean;
}

const DropdownSubListWrapper: FC = props => <StyledPopover p="0" {...props} />;

const DropdownSubList: FC<DropdownSubListProps> = ({
  disabled = false,
  title,
  children,
  ...otherProps
}) => {
  const [visible, setVisible] = useState(false);

  return (
    <Tooltip
      position={Position.RIGHT}
      visible={visible}
      mouseEnterDelay={200}
      mouseLeaveDelay={300}
      Wrapper={DropdownSubListWrapper}
      content={<List onClick={() => setVisible(false)}>{children}</List>}
      onVisibleChange={newVisible => {
        if (!disabled) {
          setVisible(newVisible);
        }
      }}
    >
      <Item disabled={disabled} {...otherProps}>
        {title}
        <Box display="inline-flex" ml="auto" pl="2">
          <Icon cursor="pointer" type={MdKeyboardArrowRight} />
        </Box>
      </Item>
    </Tooltip>
  );
};

export default DropdownSubList;
