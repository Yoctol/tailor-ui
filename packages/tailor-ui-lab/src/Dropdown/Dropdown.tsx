import React, { FunctionComponent, ReactNode } from 'react';

import { Position } from 'tailor-ui';

import Popover, { PopoverProps } from '../Popover';

import DropdownContext from './DropdownContext';
import DropdownItem from './DropdownItem';
import DropdownSubList from './DropdownSubList';
import { Divider, List } from './styles';

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

interface DropdownProps extends Omit<PopoverProps, 'content'> {
  overlay: ReactNode;
}

const Dropdown: FunctionComponent<DropdownProps> & {
  List: typeof List;
  SubList: typeof DropdownSubList;
  Item: typeof DropdownItem;
  Divider: typeof Divider;
} = ({ children, position = Position.BOTTOM_LEFT, overlay, ...props }) => {
  return (
    <Popover
      p="0"
      position={position}
      content={close => (
        <DropdownContext.Provider value={{ close }}>
          {overlay}
        </DropdownContext.Provider>
      )}
      {...props}
    >
      {children}
    </Popover>
  );
};

Dropdown.List = List;
Dropdown.SubList = DropdownSubList;
Dropdown.Item = DropdownItem;
Dropdown.Divider = Divider;

export default Dropdown;
