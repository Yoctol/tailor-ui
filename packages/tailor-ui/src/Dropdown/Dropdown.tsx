import React, { FC, ReactNode } from 'react';

import { Popover, PopoverProps } from '../Popover';
import { Position } from '../constants';

import DropdownContext from './DropdownContext';
import DropdownItem from './DropdownItem';
import DropdownSubList from './DropdownSubList';
import { Divider, List } from './styles';

export interface DropdownProps extends Omit<PopoverProps, 'content'> {
  overlay: ReactNode;
}

const Dropdown: FC<DropdownProps> & {
  List: typeof List;
  SubList: typeof DropdownSubList;
  Item: typeof DropdownItem;
  Divider: typeof Divider;
} = ({ children, position = Position.BOTTOM_LEFT, overlay, ...props }) => {
  return (
    <Popover
      p="0"
      position={position}
      content={(close) => (
        <DropdownContext.Provider value={{ close }}>
          {overlay}
        </DropdownContext.Provider>
      )}
      {...props}
    >
      {children}
    </Popover>
  );
};

Dropdown.List = List;
Dropdown.SubList = DropdownSubList;
Dropdown.Item = DropdownItem;
Dropdown.Divider = Divider;

export { Dropdown };
