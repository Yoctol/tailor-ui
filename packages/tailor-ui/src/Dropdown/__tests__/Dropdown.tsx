import React from 'react';

import { fireEvent, render, wait, waitForElement } from 'test/test-utils';

import { Dropdown } from '../Dropdown';

describe('Dropdown', () => {
  it('should render correctly', () => {
    const { baseElement } = render(
      <Dropdown
        visible
        overlay={
          <Dropdown.List>
            <Dropdown.Item>item</Dropdown.Item>
          </Dropdown.List>
        }
      >
        <button type="button">button</button>
      </Dropdown>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should not visible when click dropdown item', async () => {
    const { getByText, queryByTestId } = render(
      <Dropdown
        overlay={
          <Dropdown.List data-testid="dropdown-list">
            <Dropdown.Item>item</Dropdown.Item>
          </Dropdown.List>
        }
      >
        <button type="button">button</button>
      </Dropdown>
    );

    const button = getByText('button');

    fireEvent.click(button);

    const item = await waitForElement(() => getByText('item'));

    fireEvent.click(item);

    await wait(() =>
      expect(queryByTestId('dropdown-list')).not.toBeInTheDocument()
    );
  });

  it('should keeo visible when click dropdown keep item', async () => {
    const { getByText, queryByTestId } = render(
      <Dropdown
        overlay={
          <Dropdown.List data-testid="dropdown-list">
            <Dropdown.Item keep>item</Dropdown.Item>
          </Dropdown.List>
        }
      >
        <button type="button">button</button>
      </Dropdown>
    );

    const button = getByText('button');

    fireEvent.click(button);

    const item = await waitForElement(() => getByText('item'));

    fireEvent.click(item);

    await wait(() =>
      expect(queryByTestId('dropdown-list')).toBeInTheDocument()
    );
  });
});
