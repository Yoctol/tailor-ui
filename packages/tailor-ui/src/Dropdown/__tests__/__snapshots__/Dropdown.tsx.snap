// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  will-change: transform;
}

.c1 {
  padding: 0.5rem 1rem;
  border: 1px solid;
  border-radius: 0.5rem;
  border-color: #e0e0e0;
  background-color: #ffffff;
  box-shadow: 0 2px 4px rgba(100,120,168,0.3);
  color: #373737;
  font-size: 0.875rem;
  text-align: left;
  white-space: nowrap;
  padding: 0rem;
}

.c2 {
  max-height: 240px;
  padding: 0.5rem 0;
  overflow-x: hidden;
  overflow-y: auto;
  border-radius: 0.5rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 2.25rem;
  padding: 0 24px;
  background-color: #ffffff;
  font-size: 0.875rem;
  cursor: pointer;
}

.c3[disabled],
.c3:disabled {
  color: #8b8b8b;
  cursor: not-allowed;
}

.c3[disabled]:hover,
.c3:disabled:hover {
  background-color: #ffffff;
}

.c3:hover {
  background-color: #f6f6f6;
}

<body>
  <div>
    <button
      type="button"
    >
      button
    </button>
  </div>
  <div
    portal-container=""
  >
    <div
      style="position: fixed; top: 0px; left: 0px; z-index: 10;"
    >
      <div
        class="c0"
        style="transform: translate3d(-6px, -6px, 0px);"
      >
        <div
          style="opacity: 0; transform: scale(0.9);"
        >
          <div
            class="c1"
          >
            <div
              class="c2"
            >
              <li
                class="c3"
              >
                item
              </li>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
